parley/src/layout/line/greedy.rs

```diff
@@ -759,8 +760,11 @@ fn try_commit_line<B: Brush>(
                     cluster_range.end = state.clusters.end;
                 }
 
+                let overlap = run_data.cluster_range.contains(&cluster_range.start)
+                    && run_data.cluster_range.contains(&(cluster_range.end - 1));
                 if cluster_range.start > cluster_range.end
                     || (!is_empty && cluster_range.start == cluster_range.end)
+                    || !overlap
                 {
```

Comment:
Could this be simplified to the following?

```suggestion
                if cluster_range.start >= run_data.cluster_range.end {
```

The idea is that if our next cluster index (`state.clusters.start`) is outside of this run, we've already fully consumed the run. I think where it goes wrong with this bug is that `cluster_range.end` is also updated, not only `cluster_range.start`, because when there's a single inline box on this line, this `TextRun` is both the first and the last of this line.

Reply:
Your suggestion seems to work and the reasoning is sound, but I'm still hesitant to remove anything until I know why it was there in the first place.

So far, I figured that `(!is_empty && cluster_range.start == cluster_range.end)` is probably for mandatory line breaks, where it avoids adding an empty text run to a line. Removing it breaks the editor tests, because the cursor will be rendered in an incorrect spot, but the text is still rendered correctly.

I assume that `cluster_range.start > cluster_range.end` is an incomplete check for the bug fully fixed in this PR, but I'm not sure yet.

Reply:
I'm fairly convinced that your idea is correct, so I've committed it. As far as I can see, there is only one way to have a run with a cluster range that falls outside the line state cluster range, and that is the scenario you've described above. The `cluster_range.start >= run_data.cluster_range.end` condition is just a cleaner way to check for all possible variants of the same problem.